hikari/impl/event_manager_base.py

```diff
@@ -50,46 +56,135 @@
+
+    cache_components: undefined.UndefinedOr[config.CacheComponents]
+    """Bitfield of the cache components this consumer makes modifying calls to, if set."""
+
+    event_types: undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]]
+    """A sequence of the types of events this consumer dispatches to, if set."""
+
+
 class EventManagerBase(event_manager.EventManager):
     """Provides functionality to consume and dispatch events.
 
     Specific event handlers should be in functions named `on_xxx` where `xxx`
     is the raw event name being dispatched in lower-case.
     """
 
-    __slots__: typing.Sequence[str] = ("_app", "_listeners", "_consumers", "_waiters")
+    __slots__: typing.Sequence[str] = ("_app", "_consumers", "_enabled_consumers_cache", "_listeners", "_waiters")
 
     def __init__(self, app: traits.BotAware) -> None:
         self._app = app
-        self._consumers: typing.Dict[str, ConsumerT] = {}
+        self._consumers: typing.Dict[str, _Consumer] = {}
+        self._enabled_consumers_cache: typing.Dict[_Consumer, bool] = {}
         self._listeners: ListenerMapT[base_events.Event] = {}
         self._waiters: WaiterMapT[base_events.Event] = {}
 
         for name, member in inspect.getmembers(self):
             if name.startswith("on_"):
-                self._consumers[name[3:]] = member
+                member = typing.cast(MethodT, member)
+                cache_resource = getattr(member, _CACHE_RESOURCE_ATTRIBUTE, undefined.UNDEFINED)
+                event_types = getattr(member, _EVENT_TYPES_ATTRIBUTE, undefined.UNDEFINED)
+                cache_resource: undefined.UndefinedOr[config.CacheComponents] = cache_resource
+                event_types: undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]] = event_types
+                self._consumers[name[3:]] = _Consumer(member, cache_resource, event_types)
+
+    def _clear_enabled_cache(self) -> None:
+        self._enabled_consumers_cache = {}
+
+    def _enabled_for_event(self, event_type: typing.Type[base_events.Event], /) -> bool:
+        for cls in event_type.dispatches():
+            if cls in self._listeners or cls in self._waiters:
+                return True
+
+        return False
+
+    # This returns int rather than bool to avoid an unnecessary bool cast
+    def _enabled_for_consumer(self, consumer: _Consumer, /) -> int:
+        # If undefined then we can only assume that this may link to registered listeners.
+        if consumer.event_types is undefined.UNDEFINED:
+            return True
+
+        if (cached_value := self._enabled_consumers_cache.get(consumer)) is True:
+            return True
+
+        if cached_value is None:
+            for event_type in consumer.event_types:
+                if event_type in self._listeners or event_type in self._waiters:
+                    self._enabled_consumers_cache[consumer] = True
+                    return True
```

Comment:
Wouldn't this cause problems if the `event_type` in the waiters is removed right after this check? We will be deserializing objects that we don't need to deserialize (which is what this pr is trying to prevent).

Maybe we will have to remove the consumer entry from `_enabled_consumers_cache` when deleted off waiters or listeners

Reply:
This is a single thread environment and consume_raw_event never yields to the event loop so that'd never happen